DatadogInternal/Sources/NetworkInstrumentation/NetworkInstrumentationFeature.swift

```diff
@@ -49,80 +52,72 @@ internal final class NetworkInstrumentationFeature: DatadogFeature {
     /// - Parameter configuration: The configuration to use for swizzling.
     /// Note: We are only concerned with type of the delegate here but to provide compile time safety, we
     ///      use the instance of the delegate to get the type.
-    internal func bindIfNeeded(configuration: URLSessionInstrumentation.Configuration) throws {
+    internal func bind(configuration: URLSessionInstrumentation.Configuration) throws {
         let configuredFirstPartyHosts = FirstPartyHosts(firstPartyHosts: configuration.firstPartyHostsTracing) ?? .init()
 
-        try URLSessionTaskDelegateSwizzler.bindIfNeeded(
-            delegateClass: configuration.delegateClass,
-            interceptDidFinishCollecting: { [weak self] session, task, metrics in
-                self?.queue.async { [weak self, weak session] in
-                    self?._task(task, didFinishCollecting: metrics)
-                    session?.delegate?.interceptor?.task(task, didFinishCollecting: metrics)
+        let identifier = ObjectIdentifier(configuration.delegateClass)
 
-                    // iOS 16 and above, didCompleteWithError is not called hence we use task state to detect task completion
-                    // while prior to iOS 15, task state doesn't change to completed hence we use didCompleteWithError to detect task completion
-                    if #available(iOS 15, tvOS 15, *) {
-                        self?._task(task, didCompleteWithError: task.error)
-                        session?.delegate?.interceptor?.task(task, didCompleteWithError: task.error)
-                    }
+        if let swizzler = swizzlers[identifier] {
+            DD.logger.warn(
+                """
+                The delegate class \(configuration.delegateClass) is already instrumented.
+                The previous instrumentation will be disabled in favor of the new one.
+                """
+            )
+
+            swizzler.unswizzle()
+        }
+
+        let swizzler = NetworkInstrumentationSwizzler()
+        swizzlers[identifier] = swizzler
+
+        try swizzler.swizzle(
+            interceptResume: { [weak self] task in
+                // intercept task if delegate match
+                guard let self = self, task.dd.delegate?.isKind(of: configuration.delegateClass) == true else {
+                    return
                 }
```

Comment:
How is the dd.delegate lifecycle is managed?

Reply:
It's a computed `var` and not retained. see its implementation [here](https://github.com/DataDog/dd-sdk-ios/blob/738d599b46f982275a6c41c48578808403f36eb6/DatadogInternal/Sources/NetworkInstrumentation/URLSession/URLSessionTask%2BTracking.swift#L23L35)